lldp_ethtool.c

```diff
@@ -0,0 +1,296 @@
+		   nlhdr->nlmsg_len, nlhdr->nlmsg_type, ghdr->cmd);
+
+	return NL_STOP;
+}
+
+static int ethtool_msg_mm_get_parse(struct nl_msg *nlm, void *arg)
+{
+	struct genlmsghdr *hdr = nlmsg_data(nlmsg_hdr(nlm));
+	struct nlattr *tb[ETHTOOL_A_MM_MAX];
+	struct ethtool_mm *mm = arg;
+
+	if (nla_parse(tb, ETHTOOL_A_MM_MAX, genlmsg_attrdata(hdr, 0),
+		      genlmsg_attrlen(hdr, 0), ethtool_mm_genl_policy)) {
+		printf("nla_parse() failed\n");
+		return NL_STOP;
+	}
+
+	if (tb[ETHTOOL_A_MM_VERIFY_STATUS])
+		mm->verify_status = nla_get_u8(tb[ETHTOOL_A_MM_VERIFY_STATUS]);
+
+	if (tb[ETHTOOL_A_MM_VERIFY_TIME])
+		mm->verify_time = nla_get_u32(tb[ETHTOOL_A_MM_VERIFY_TIME]);
+
+	if (tb[ETHTOOL_A_MM_MAX_VERIFY_TIME])
+		mm->max_verify_time =
+			nla_get_u32(tb[ETHTOOL_A_MM_MAX_VERIFY_TIME]);
+
+	if (tb[ETHTOOL_A_MM_TX_ENABLED])
+		mm->tx_enabled = nla_get_u8(tb[ETHTOOL_A_MM_TX_ENABLED]);
+
+	if (tb[ETHTOOL_A_MM_TX_ACTIVE])
+		mm->tx_active = nla_get_u8(tb[ETHTOOL_A_MM_TX_ACTIVE]);
+
+	if (tb[ETHTOOL_A_MM_PMAC_ENABLED])
+		mm->pmac_enabled = nla_get_u8(tb[ETHTOOL_A_MM_PMAC_ENABLED]);
+
+	if (tb[ETHTOOL_A_MM_TX_MIN_FRAG_SIZE])
+		mm->tx_min_frag_size = nla_get_u32(tb[ETHTOOL_A_MM_TX_MIN_FRAG_SIZE]);
+
+	if (tb[ETHTOOL_A_MM_RX_MIN_FRAG_SIZE])
+		mm->rx_min_frag_size = nla_get_u32(tb[ETHTOOL_A_MM_RX_MIN_FRAG_SIZE]);
+
+	return NL_OK;
+}
+
+static int ethtool_genl_start(int cmd, int header_type, const char *ifname,
+			      struct nl_sock **sock, struct nl_msg **msg)
+{
+	struct nlmsghdr *hdr;
+	struct nlattr *attr;
+	struct nl_sock *sk;
+	struct nl_msg *nlm;
+	int err, family_id;
+
+	sk = nl_socket_alloc();
+	if (!sk) {
+		LLDPAD_ERR("Failed to allocate netlink socket\n");
+		return -ENOMEM;
+	}
+
+	err = genl_connect(sk);
+	if (err) {
+		LLDPAD_ERR("Failed to connect to netlink socket\n");
+		goto out_free;
+	}
+
+	err = genl_ctrl_resolve(sk, ETHTOOL_GENL_NAME);
+	if (err < 0) {
+		LLDPAD_ERR("Failed to resolve ethtool genl family id\n");
+		goto out_close;
+	}
+
+	family_id = err;
+
+	nlm = nlmsg_alloc();
+	if (!nlm) {
+		LLDPAD_ERR("Failed to allocate netlink message\n");
+		err = -ENOMEM;
+		goto out_close;
+	}
+
+	hdr = genlmsg_put(nlm, NL_AUTO_PORT, NL_AUTO_SEQ, family_id, 0,
+			  NLM_F_REQUEST | NLM_F_ACK, cmd, ETHTOOL_GENL_VERSION);
+	if (!hdr) {
+		LLDPAD_ERR("genlmsg_put() failed\n");
+		err = -EMSGSIZE;
+		goto out_free_nlmsg;
+	}
+
+	attr = nla_nest_start(nlm, header_type);
+	nla_put_string(nlm, ETHTOOL_A_HEADER_DEV_NAME, ifname);
+	nla_nest_end(nlm, attr);
+
+	*sock = sk;
+	*msg = nlm;
+
+	return 0;
+
+out_free_nlmsg:
+	nlmsg_free(nlm);
+out_close:
+	nl_close(sk);
+out_free:
+	nl_socket_free(sk);
+	return err;
+}
+
+static void ethtool_genl_stop(struct nl_sock *sk, struct nl_msg *nlm)
+{
+	nlmsg_free(nlm);
+	nl_close(sk);
+	nl_socket_free(sk);
+}
+
+int ethtool_mm_get_state(const char *ifname, struct ethtool_mm *mm)
+{
+	struct nl_sock *sk;
+	struct nl_msg *nlm;
+	int err;
+
+	err = ethtool_genl_start(ETHTOOL_MSG_MM_GET, ETHTOOL_A_MM_HEADER,
```

Comment:
Is there a reason to keep creating / deleting the netlink socket?

Reply:
I guess I didn't want an API that exposes a stateful struct ethtool_sock *, because that would be required with a reusable socket. I'll try and add it and see how it goes.